orchagent/neighorch.h

```diff
@@ -67,6 +69,17 @@ class NeighOrch : public Orch, public Subject
     bool clearNextHopFlag(const NextHopKey &, const uint32_t);
 
     void doTask(Consumer &consumer);
+    void doVoqSystemNeighTask(Consumer &consumer);
+
+    unique_ptr<Table> m_tableVoqSystemNeighTable;
+    bool getSystemPortNeighEncapIndex(string alias, uint32_t &encap_index);
+    bool addVoqEncapIndex(string &alias, IpAddress &ip, vector<sai_attribute_t> &neighbor_attrs);
+    bool addKernelRoute(string odev, IpAddress ip_addr);
+    bool delKernelRoute(IpAddress ip_addr);
+    bool addKernelNeigh(string odev, IpAddress ip_addr, MacAddress mac_addr);
```

Comment:
So far, orchagent hasn't directly configured anything to kernel. This is a behavioral change.

Reply:
Agree, in terms of ARP/Route, this is the exception and discussed in HLD, to keep VOQ specific changes to be simple enough.

Reply:
Can you explain why nbrmgr handle this scenario?

Reply:
remote neighbors are created in CHASSIS_APP_DB ("app_db") and hence its processed inside OrchAgent. Now, OrchAgent could post another message to nbrmgr as if its config_db event, but its adding many indirection instead of syncing to ASIC_DB.

Reply:
Another proposal is `CHASSIS_APP_DB -> Nbrmgr ->(Kernel, APP_DB) -> orchagent`. As we know that orchagent currently don't handle kernel programming, another issue with this model is related to warmboot reconciliation logic. We dont want to introduce orchagent to deal with reconciling kernel objects. Once orchagent gets the data from APP_DB, it can use `getSystemPortNeighEncapIndex `to get the EncapIndex.  Kindly revisit this design.

Reply:
Ok. We will go with nbrmgr for adding kernel ARP entry and static route. Adding static route from nbrmgr isn't clean either because nbrmgr doesnt deal with static routes as overall. Are we Ok to put static route addition in nbrmgr?.

Reply:
Ok. We will go with nbrmgr for adding static ARP entry in kernel and also adding static route. Certainly nbrmgr isn't right place for adding static route but we dont want to create another daemon for adding static routes alone. is it ok put static routes adds from nbrmgr?. Any thoughts?

Reply:
IMO, it is ok to have nbrmgr to configure static route.. I think it is for the same neighbor entry that you are creating the static route as well, right?

Reply:
That's right. its static route(/32 or /128) the neighbor entry being added.

Reply:
With new design:

chassis_app_db (remote neighbors)----> nbrmgr ---> create kernel entries(route & arp) ---> nbrsyncd ---> app_db (local)
chassis_app_db (remote neighbors)----> orchagent ---> create sai neighbor for remote neighbor with encap-id.... all dependencies are resolved (system-port, system routed interface). 
app_db (local neighbor entry created for remote neighbor) ---> orchagent ---> mark it as already programmed. 

when orchagent reads the local app_db neighbor object, it will map it existing remote neighbor because its already programmed and this way its sync's remote neighbor entries faster.  In case of orchagent slow in reading chassis_app_db, it will mark local neighbor entry as unresolved and wait for remote neighbor entry with encap id to appear to program the neighbor. 

kernel programming is make to iBGP routes to be appear with actual nexthop. 

We wanted clarify the new approach here. Please let us know if your comments/views on this.

Reply:
Only one subscriber for chassis_app_db. Orchagent can read that for getting the encap_id. For chassis_app_db/app_db, there can be only one subscriber.

Reply:
Do you mean, 

OrchAgent is a consumer of chassis_app_db
NbrMgr is a subscriber of chassis_app_db.

Reply:
Guohan's proposal: Let the orchagent program SAI for remote neighbor independently. Set state in state DB or local APP DB with mac address an additional attribute.. Nbrmgr programs kernel from the STATE DB/App DB entry only after making sure that SAI is programmed for the remote neighbors.  Nbrmgr will not use the chassis_app_db.

Reply:
Fixed. Implemented the proposal with STATE_DB to have the entries to indicate SAI program completion. Orchagent will not program the kernel. Orchagent programs only SAI.